R/utilities-ratio-comparison.R

```diff
@@ -248,92 +245,161 @@ getPKRatioSummaryFromAnalyticalSolution <- function(pkData, referencePKData, sim
 }
 
 
-#' @title getPKRatioSummaryFromMCSampling
+#' @title getPKRatioSummaryFromSettings
 #' @description Get PK Parameter Ratio Measure From Monte Carlo Sampling
 #' @param pkData A data.frame of PK Parameter values for Population to compare
 #' @param referencePKData A data.frame of PK Parameter values for reference Population
-#' @param simulationSetName Name of simulation set
 #' @param settings A list of task settings
 #' @return A data.frame of the PK Parameter ratios summary statistics
 #' @keywords internal
-getPKRatioSummaryFromMCSampling <- function(pkData, referencePKData, simulationSetName, settings = NULL) {
+getPKRatioSummaryFromSettings <- function(pkData, referencePKData, settings) {
   # Sample from largest population if size is different
   # Create a list of Sampled PK Parameters for each MC repetition and calculate their Ratio
   pkSize <- length(unique(pkData$IndividualId))
   referenceSize <- length(unique(referencePKData$IndividualId))
   mcRepetitions <- settings$mcRepetitions %||% getDefaultMCRepetitions()
   mcRandomSeed <- settings$mcRandomSeed %||% getDefaultMCRandomSeed()
 
-  t0 <- tic()
-  logInfo(messages$monteCarlo(simulationSetName, mcRepetitions, mcRandomSeed))
+  widerPopulation <- ifelse(pkSize < referenceSize, "reference", "comparison")
+
+  selectedIds <- mcSampling(
+    dataSize = switch(widerPopulation,
+      "reference" = referenceSize,
+      "comparison" = pkSize
+    ),
+    sampleSize = switch(widerPopulation,
+      "reference" = pkSize,
+      "comparison" = referenceSize
+    ),
+    n = mcRepetitions,
+    seed = mcRandomSeed
+  )
+  useParallel <- all(
+    requireNamespace("parallel", quietly = TRUE),
+    settings$numberOfCores > 1
```

Comment:
if I set `workflow$calculatePKParameters$settings$numberOfCores` to the number of available cores on my machine, the execution time becomes 4.9 minutes - which is still slightly slower than 4.7 minutes before the fix.

Reply:
The test case scenario is as following:
* 8 PK parameters:
```
pkParameters <-
  c('C_max', 'C_max_norm', 't_max', 'C_tEnd', 'AUC_tEnd', 'AUC_tEnd_norm', 'AUC_inf', 'AUC_inf_norm')
```
* 2 simulation sets
  * 1000 individuals in the reference population
  * 1000 individuals in the non-reference population
* 1 Output per simulation set

Reply:
![grafik](https://github.com/Open-Systems-Pharmacology/OSPSuite.ReportingEngine/assets/25061876/9ae90849-9d25-487c-bdfc-9090e135e530) 
Also wondering that the parallelization factor is less than 3 with 8 cores...

Reply:
> if I set `workflow$calculatePKParameters$settings$numberOfCores` to the number of available cores on my machine (8), the execution time becomes 4.9 minutes - which is still slightly slower than 4.7 minutes before the fix.

Yes, it is the correct settings for the number of cores

Reply:
> Also wondering that the parallelization factor is less than 3 with 8 cores... And I can see that 8 processes are indeed created in parallel.

Only the Monte Carlo sampling portion is parallelized, so all the set up and PK parameter plots time are not shortened.
Additionally, setting up the parallel environment with a big data.frame of 8000 rows may also add to the computation time. "However, it is rare that we would achieve this optimal speed-up since there is always communication between threads" (from https://csgillespie.github.io/efficientR/performance.html)

Reply:
@Yuri05 
I just updated to another method which should make things much faster even without parallelization.
Briefly, before the calculation I pivoted the data.frame and translated it into a matrix object.
The benefit from numeric matrix objects is that computations can be vectorized.

If the build pass, let me know if this updated method works as expected and is efficient.